WalletWasabi/BranchNBound/BranchAndBound.cs

```diff
@@ -0,0 +1,178 @@
+using NBitcoin;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.BranchNBound
+{
+	/// <seealso href="https://murch.one/wp-content/uploads/2016/11/erhardt2016coinselection.pdf">Section "5.3 Branch and Bound".</seealso>
+	public class BranchAndBound
+	{
+		private readonly Random _random = new();
+
+		/// <param name="values">All values must be strictly positive.</param>
+		public BranchAndBound(List<long> values)
+		{
+			Count = values.Count;
+			SortedValues = values.OrderByDescending(x => x).ToArray();
+
+			foreach (var value in SortedValues)
+			{
+				if (value <= 0)
+				{
+					throw new ArgumentException("Only strictly positive values are supported.");
+				}
+			}
```

Comment:
Can we simplify this check?

```suggestion
				if (values.Any(x => x <= 0))
				{
					throw new ArgumentException("Only strictly positive values are supported.");
				}
			}
```

Reply:
Or, because it's sorted **descending**, just check the last item in the array.
If it's positive, all the other items will be positive as well for sure.

Reply:
The array might also be empty. I have modified it a bit.